packages/hardware-ledger/src/LedgerKeyAgent.ts

```diff
@@ -148,52 +148,101 @@ export class LedgerKeyAgent extends KeyAgentBase {
     this.#communicationType = serializableData.communicationType;
   }
 
-  static findKeyAgentByCommunicationTypeAndDevicePath(communicationType: CommunicationType, devicePath?: string) {
-    return this.deviceConnections?.find(
+  private static async findConnectionByCommunicationTypeAndDevicePath(
+    communicationType: CommunicationType,
+    devicePath?: string
+  ): Promise<LedgerConnection | null> {
+    const matchingConnectionData = this.deviceConnections?.find(
       (connection) => connection.communicationType === communicationType && connection.devicePath === devicePath
     );
+    if (!matchingConnectionData) return null;
+
+    try {
+      await this.testConnection(matchingConnectionData.deviceConnection);
+    } catch (error) {
+      if (
+        error instanceof errors.TransportError &&
+        error.message.includes('Cannot communicate with Ledger Cardano App')
+      ) {
+        this.deviceConnections = this.deviceConnections.filter(
+          (connectionData) => connectionData !== matchingConnectionData
+        );
+        return null;
+      }
+
+      throw error;
+    }
+
+    return matchingConnectionData.deviceConnection;
   }
 
   /**
    * @throws TransportError
    */
-  static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
+  private static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
     try {
-      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebHID.list();
+      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebUSB.list();
     } catch (error) {
       throw new errors.TransportError('Cannot fetch device list', error);
     }
   }
 
+  private static attachDisconnectionCleanupHandler(transport: LedgerTransportType) {
+    const onDisconnect = () => {
+      this.deviceConnections = this.deviceConnections.filter(
+        ({ deviceConnection }) => deviceConnection.transport !== transport
+      );
+      transport.close();
```

Comment:
This is an async function, linter complains here. Can it fail? Also, do we need to clean up the event listener?
```suggestion
      void transport.close();
      transport.off('disconnect', onDisconnect);
```

Reply:
We shoul be careful here. I think you will get a disconnect event even when you open or close Cardano App since the device path will change. Could you check that @szymonmaslowski  🙏

Reply:
@mkazlauskas "Can it fail?" - good question. I checked the underlying implementation and the ledger's code cannot fail, but it calls the USBDevice API, which is a black box. There is another occurrence of closing the transport [here](https://github.com/input-output-hk/cardano-js-sdk/blob/514df4872b4cfe0364cba676e34f80b32ddfa7f2/packages/hardware-ledger/src/LedgerKeyAgent.ts#L310) which was implemented a long time ago and it doesn't have error handling. However, IMO the safest way would be to add error handling in both places with a simple error log.

@tomislavhoracek Yes the disconnect event will happen also on closing the cardano app. This is all fine IMO. Actually, this is the reason for this code. Do you remember the page reload on the lace side happening on error to the connection? I guess it was introduced because the connection stayed remembered and when you closed the cardano app or disconnected the cable and tried to reconnect the remembered connection was returned but that connection was already stale so lace was stuck and couldn't connect. At first, I wanted to remove this capability of remembering connections but it makes more sense to test the connection and eventually forget it. And when you forget it makes sense to close it because it won't be used.